### ISSUE TITLE
Built-in Python can have lib64 issues

### ISSUE THREAD

**Issue by amundson** [spack/spack#7353]
I have identified a problem building Python modules against the system Python on systems that use lib64 instead of lib, i.e., RedHat and derivatives. Other lib64-using distributions, including SuSe, probably have the same issue.

I have the following in packages.yaml:
```
packages:
    python:
        paths:
            python@2.7.5: /usr
```
If I then build, e.g., py-mpi4py, the module files are installed in <prefix>/lib64/python2.7/site-packages. However, spack adds <prefix/lib/python2.7/site-packages to my PYTHONPATH, so use of py-mpi4py fails.

I *almost* have a suggested change to PythonPackage in python.py. 
If I change
```python
    def install_args(self, spec, prefix):
        """Arguments to pass to install."""
        args = ['--prefix={0}'.format(prefix)]
```
to
```python
    def install_args(self, spec, prefix):
        """Arguments to pass to install."""
        args = ['--prefix={0}'.format(prefix)]
        args += ['--install-lib={0}'.format(join_path(prefix, 'lib',
                                            'python2.7', 'site-packages'))]
```
the problem is solved, but I have hardcoded the version "2.7" for python. I don't see how to get the correct python version. If someone can help me, I will submit a pull request.

**Comment by adamjstewart** [comment]
Why would a change to `--install-lib` affect the pre-installed system Python?

**Comment by amundson** [comment]
I guess I didn't explain that very clearly. Here's a concrete example:

I have two versions of python installed. 2.7.5 is the (redhat) system python and is listed in packages.yaml. 2.7.14 was installed via "spack install python@2.7.14".

Using the spack-installed python 2.7.14 works as expected.
```
spack install py-mpi4py^python@2.7.14
```
puts the mpi4py package in
```
opt/spack/linux-rhel7-x86_64/gcc-4.8.5/
py-mpi4py-2.0.0-dm2etp3dxnptxuv3rojuqeb7vld2mk2f/lib/python2.7/site-packages
```
I can then do "spack load -r py-mpi4py^python@2.7.5", and spack will add the appropriate directory to my PYTHONPATH.

If, instead, I use the system python, 2.7.5,
```
spack install py-mpi4py^python@2.7.5
```
the mpi4py package appears in
```
opt/spack/linux-rhel7-x86_64/gcc-4.8.5/
py-mpi4py-2.0.0-zuez6rfgnb4abjbyjempjul7tymoldyw/lib64/python2.7/site-packages
```
Note the "lib64" part of the path. Redhat modifies the system python's distutils to put compiled code in a subdirectory of lib64. Pure-python libraries still use lib. My py-mpi4py^python@2.7.5 can't be used with spack because the spack-generated PYTHONPATH does not include the lib64-derived directory.

These problems can be avoided if the spack PythonPackage tells distutils to use the plain "lib" directory. My code snippet at the top does that, but it hardcodes "2.7" for the python version because I could not figure out how to derive the python version in that context.

**Comment by adamjstewart** [comment]
Ooh, very interesting. I think I see what you're saying now. We shouldn't be making changes to `python/package.py`, we should be making changes to `PythonPackage` so that every Python module is installed with an explicit lib directory. I would support this change. Seems like the right way to solve it to me.

**Comment by adamjstewart** [comment]
Alternatively, we could add `prefix/lib64/python2.7/site-packages` to the `PYTHONPATH`. It might be better to keep the installations simple for the sake of being able to `spack activate` things.

**Comment by jorblancoa** [comment]
I have the exact same problem. I was trying to solve it by creating symlinks from lib64 to lib, but I guess your solution is cleaner.

To get the python version you can use:
`spec['python'].version.up_to(2)`

**Comment by chuckatkins** [comment]
Rather than explicitly adding lib and lib64, python itself can be queried for the correct values with `distutils.sysconfig.get_python_lib(plat_specific={True,False})`:
```
$ python3 -c 'import distutils.sysconfig; print(distutils.sysconfig.get_python_lib(plat_specific=True)); print(distutils.sysconfig.get_python_lib(plat_specific=False))'
/usr/lib64/python3.6/site-packages
/usr/lib/python3.6/site-packages
```

**Comment by adamjstewart** [comment]
Looking back at this, I don't think the solution is to force Spack to install libraries to `lib`, since pre-installed system libraries may already be present in `lib64`. I think the right solution is to check whether libraries are installed in `lib` or `lib64` as @chuckatkins proposed. We already do this in the `libs` property of the `python` package, but I guess we don't do it for `site_packages_dir`. Shouldn't be hard to modify if someone wants to take a stab at it.

**Comment by adamjstewart** [comment]
This should be fixed by #28346. We now query `sysconfig.get_paths` to determine the correct purelib (`lib/pythonX.Y/site-packages`) and platlib (`lib64/pythonX.Y/site-packages`) and add both to the `PYTHONPATH`.
